experiments/prediction/figure1.py

```diff
@@ -0,0 +1,158 @@
+        # go LEFT 40% of the time
+        'target': actionArrayToPolicy([0.4, 0.6]),
+        # take each action equally
+        'behavior': actionArrayToPolicy([0.5, 0.5]),
+        'gamma': 1.0,
+        'steps': 3000,
+        # hardcode stepsizes found from parameter study
+        'stepsizes': {
+            'TD': 0.01,
+        }
+    },
+    # 5-state random walk environment with inverted features
+    {
+        'env': RandomWalk,
+        'representation': Inverted,
+        # go LEFT 40% of the time
+        'target': actionArrayToPolicy([0.4, 0.6]),
+        # take each action equally
+        'behavior': actionArrayToPolicy([0.5, 0.5]),
+        'gamma': 1.0,
+        'steps': 3000,
+        # hardcode stepsizes found from parameter study
+        'stepsizes': {
+            'TD': 0.01,
+        }
+    },
+]
+
+COLORS = {
+    'TD': 'blue',
+}
+
+# -----------------------------------
+# Collect the data for the experiment
+# -----------------------------------
+
+# a convenience object to store data collected during runs
+collector = Collector()
+
+for run in range(RUNS):
+    for problem in PROBLEMS:
+        for Learner in LEARNERS:
+            # for reproducibility, set the random seed for each run
+            # also reset the seed for each learner, so we guarantee each sees the same data
+            np.random.seed(run)
+
+            # build a new instance of the environment each time
+            # just to be sure we don't bleed one learner into the next
+            Env = problem['env']
+            env = Env()
+
+            target = problem['target']
+            behavior = problem['behavior']
+
+            Rep = problem['representation']
+            rep = Rep()
+
+            # build the X, P, R, and D matrices for computing RMSPBE
+            X, P, R, D = env.getXPRD(target, rep)
+            RMSPBE = buildRMSPBE(X, P, R, D, problem['gamma'])
+
+            # build a new instance of the learning algorithm
+            learner = Learner(rep.features(), {
+                'gamma': problem['gamma'],
+                'alpha': problem['stepsizes'][Learner.__name__]
+            })
+
+            # build an "agent" which selects actions according to the behavior
+            # and tries to estimate according to the target policy
+            agent = RlGlueCompatWrapper(learner, behavior, target, rep.encode)
+
+            # build the experiment runner
+            glue = RlGlue(agent, env)
+
+            # start the episode
+            glue.start()
+            for step in range(problem['steps']):
+                _, _, _, terminal = glue.step()
+
+                # when we hit a terminal state, start a new episode
+                if terminal:
+                    glue.start()
+
+                # evaluate the RMPSBE
+                w = learner.getWeights()
+                rmspbe = RMSPBE(w)
+
+                collector.collect(f'{Env.__name__}-{Rep.__name__}-{Learner.__name__}', rmspbe)
```

Comment:
I'm not sure if I understand what this does so it might not be instantly clear to the reader either.

Reply:
Yup I'll add a comment here